proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,2290 @@
+unrestricted function is said to have multiplicity ``ω``. Multiplicity
+polymorphic functions may have variable multiplicity (see also Syntax_), *e.g.*
+
+::
+
+  map :: (a :p-> b) -> [a] :p-> [b]
+
+Without polymorphism, we would need two implementations of ``map`` with
+the exact same code: one for ``p=1`` and one for ``p=ω``. Function
+composition is even worse: it takes two multiplicity parameters,
+hence, would require four identical implementations:
+
+::
+
+  (.) :: (b :p-> c) -> (a :q-> b) -> a :(p ':* q)-> c
+
+Syntax
+------
+
+.. _Syntax:
+
+*This section is only a proposal for a syntax for linear types. We had
+to choose a syntax for the purpose of the proposal, but we are not
+particularly decided on it. We have listed all the alternatives in the
+Alternatives_ section.*
+
+This proposal adds two new syntactical constructs:
+
+- The multiplicity annotated arrow, for polymorphism, is written
+  ``a :p-> b`` (where ``a`` and ``b`` are types and ``p`` is a
+  multiplicity). To avoid introducing a new notion of "mixfix"
+  operators, we introduce a familly of (infix) type constructors:
+  ``(:p->)`` for each multiplicity ``p``. This technically steals
+  syntax as ``(:)`` is a valid type operator under the discouraged
+  ``-XDataKinds`` syntax. But this should not be a problem in
+  practice.
+
+  - In ``(:p->)``, ``p`` can be any type expression of kind
+    ``Multiplicity`` (see below). So that the following is legal
+    (though see Alternatives_):
+
+    ::
+
+      type familly F (a :: *) :: Multiplicity
+      f ::  forall (a :: *). Int  :(F a)-> a -> a
+- Binders can also be annotated with a multiplicity:
+
+  ::
+
+    \x ::('One) A -> x
+
+  is the identity function at type ``A ->. A``. This adds a new syntactic for binders:
+
+  ::
+
+    VAR ::( TYPE_EXPRESSION ) TYPE_EXPRESSION
+
+  where the first ``TYPE_EXPRESSION`` must be of kind ``Multiplicity``
+  (see below). This creates a new lexical token ``::(``, which does
+  not seem to be a problem.
+
+  This form is disallowed for:
+
+  - Type variables
+
+    ::
+
+      forall (a ::('One)). a -> Int -- rejected
+  - Top-level signatures (though, see `Toplevel binders`_)
+
+    ::
+
+      foo ::('One) A -> B -- rejected
+      foo x = …
+
+  The form is however permitted in records (see `Records`_ below)
+
+  ::
+
+    data R = R { unrestrictedField ::(Omega) A, linearField ::(One) B }
+
+In the fashion of levity polymorphism, the proposal introduces a data
+type ``Multiplicity`` which is treated specially by the type checker,
+to represent the multiplicities:
+
+- ::
+
+    data Multiplicity
+      = One    -- represents 1
+      | Omega  -- represents ω
+
+- Accompanied by two specially recognised type families:
+
+  ::
+
+    type family (:+) (p :: Multiplicity) (q :: Multiplicity) :: Multiplicity
+    type family (:*) (p :: Multiplicity) (q :: Multiplicity) :: Multiplicity
+
+  Note: unification of
+  multiplicities will be performed up to the semiring laws for
```

Comment:
What is the status of this line of the proposal? There has been some conversation that this unification algorithm is undecidable. Do you still intend to implement it? (Other aspects of GHC type inference are also undecidable -- at least with `UndecidableInstances` -- so that's not a problem, per se.)

Reply:
Without experience in this area, it's hard for me to understand this all without examples. But I wouldn't respond here -- please put the answer in the proposal itself. Given that we're not yet sure about what will be feasible, it might be nice to have a description of the ideal scenario and a sure-we-can-do-it fallback scenario (like your last paragraph, above). In that text, can you include an example of what's "dreadfully unhelpful"?

Reply:
I've tried to clarified the Unresolved questions section with the outcome of this conversation. I'd like to leave it at this until I've had time to actually propose a design for inference. As I'd like to avoid doing research in this proposal.